New ingester module: Envoy access-log-server (#37)
seznam/slo-exporter: docs/modules/envoy_access_log_server.md

# Envoy access log server

|                |                        |
|----------------|------------------------|
| `moduleName`   | `envoyAccessLogServer` |
| Module type    | `producer`             |
| Output event   | `raw`                  |

This module starts a grpcServer ready to accept envoy access logs (see below for supported API version) - it implements [envoy's gRPC ALS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/accesslog/v3/als.proto#envoy-v3-api-field-service-accesslog-v3-streamaccesslogsmessage-http-logs).

### Envoy support and configuration

> At this moment, V3 of envoy's xDS API is supported. [See the upstream documentation for details](https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versio) on API versions.

Reviewer: ```suggestion
At this moment, V3 of envoy's xDS API is supported. [See the upstream documentation for details](https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versions) on API versions.
```

See a minimal example on how to configure envoy to send access_logs to an slo-exporter instance (envoy v1.15+):
```
static_resources:

  listeners:
  - name: listener_0
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 8080
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          access_log:
          - name: envoy.access_loggers.http_grpc
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.access_loggers.grpc.v3.HttpGrpcAccessLogConfig
              common_config:
                grpc_service:
                  envoy_grpc:
                    cluster_name: service_accesslog
                buffer_size_bytes:
                  value: 0
                log_name: accesslogv3
                transport_api_version: V3 # needed to ensure that v3.AccessLogService is used

>               additional_request_headers_to_log: ['X-TEST', 'SOMEOHTER-HEADER']

Reviewer: maybe (e.g. not mandatory) would be nice to show real world exmaple headers, e.g. as 

```suggestion
              additional_request_headers_to_log: ['slo-result', 'slo-class', 'slo-app', 'slo-endponit']
```

          http_filters: [...]
  clusters:
  - name: service_accesslog
    connect_timeout: 6s
    type: LOGICAL_DNS
    load_assignment:
      cluster_name: service_accesslog
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: localhost
                port_value: 18090
    http2_protocol_options: {}
    ...
```

### Resulting event's metadata content
#### Common properties

> * downstreamDirectRemoteAddress - e.g. `77.75.74.172`

Author: Table it is.

f1c3078

* downstreamDirectRemotePort - e.g. `443`
* downstreamLocalAddress - e.g. `77.75.74.172`
* downstreamLocalPort - e.g. `443`
* downstreamRemoteAddress - e.g. `77.75.74.172`
* downstreamRemotePort - e.g. `443`
* routeName
* sampleRate - e.g. `0`, `1`, `0.5`
* startTime - in RFC3339, thus e.g. `2020-12-22T14:27:28Z`
* timeToFirstDownstreamTxByte - in ns, e.g. `32451342`
* timeToFirstUpstreamRxByte - in ns, e.g. `32451342`
* timeToFirstUpstreamTxByte - in ns, e.g. `32451342`
* timeToLastDownstreamTxByte - in ns, e.g. `32451342`
* timeToLastRxByte - in ns, e.g. `32451342`
* timeToLastUpstreamRxByte - in ns, e.g. `32451342`
* timeToLastUpstreamTxByte - in ns, e.g. `32451342`
* upstreamCluster
* upstreamLocalAddress - e.g. `77.75.74.172`
* upstreamLocalPort - e.g. `443`
* upstreamRemoteAddress - e.g. `77.75.74.172`
* upstreamRemotePort - e.g. `443`
* upstreamTransportFailureReason

#### HTTP access_log request properties

> * authority

Author: f1c3078

* forwardedFor
* originalPath
* path
* port
* referer
* http_*request_header_name* (e.g. `http_slo-domain`)
* requestBodyBytes
* requestHeadersBytes
* requestId
* requestMethod
* scheme
* userAgent

#### HTTP access_log response properties
* responseBodyBytes
* responseCodeDetails

> * responseCode

Author: agreed within offline discussion that no such prefixes will be added.

Documentation specifically described what is the content of each single metadata key.

# Envoy access log server

|                |                        |
|----------------|------------------------|
| `moduleName`   | `envoyAccessLogServer` |
| Module type    | `producer`             |
| Output event   | `raw`                  |

> This module starts a grpcServer ready to accept envoy access logs (see below for supported API version)	|

Reviewer 2: ```suggestion
This module allows you to generate events based on access logs sent form remote [Envoy proxy](https://www.envoyproxy.io/) over a gRPC interface. 
```

Author: b4b0585

### Envoy support and configuration
At this moment, V3 of envoy's xDS API is supported. [See the upstream documentation for details](https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versions) on API versions.

In particular, you have to configure your envoy instance to send access_logs using [v3 AccessLogService rpc](https://github.com/envoyproxy/envoy/blob/842485709d651a6057b2ffb505ffced21173e004/api/envoy/service/accesslog/v3/als.proto#L39). We currently do not implement handling of other versions.

> 

Reviewer 2: ```suggestion

> Please note that Envoy [does not buffer any of the messages sent](https://godoc.org/github.com/envoyproxy/go-control-plane/envoy/service/accesslog/v3) so there may be possible los of the log messages.

```

Author: I would say this is misleading. Envoy allows buffering to be configured.

[doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/grpc/v3/als.proto#envoy-v3-api-msg-extensions-access-loggers-grpc-v3-commongrpcaccesslogconfig)

See a minimal example on how to configure envoy to send access_logs to an slo-exporter instance (envoy v1.15+):

> ```

Reviewer 2: ```suggestion
```yaml
```

Author: b4b0585

        port_value: 8080
    filter_chains:
   	|
     	|
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          access_log:
         	|
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.access_loggers.grpc.v3.HttpGrpcAccessLogConfig
              common_config:
                grpc_service:
                  envoy_grpc:
                    cluster_name: service_accesslog
                buffer_size_bytes:
                  value: 0
                log_name: accesslogv3
                transport_api_version: V3 # needed to ensure that v3.AccessLogService is used
              additional_request_headers_to_log: ['slo-result', 'slo-class', 'slo-app', 'slo-endpoint']
          http_filters: [...]
  clusters:
 	|
    connect_timeout: 6s
    type: LOGICAL_DNS
    load_assignment:
      cluster_name: service_accesslog
      endpoints:
     	|
       	|
            address:
              socket_address:
                address: localhost
                port_value: 18090
    http2_protocol_options: {}
    ...
```

> ### Resulting event's metadata content

Reviewer 2: ```suggestion
### Resulting event metadata
```

Author: b4b0585

> Please note that some of the keys may not be present	|

Author: In certain cases we do not add keys for which we were not able to process the received data:
* Downstream/Upstream addresses - if we are not able to process them as a `SocketAddress` type
* Timestamps

Because of that, I would leave the original sentence.

#### Common properties

> | metadata's key                    |  example(s)            | description |

Author: I prefer the original.

*Note: please see [envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage) on explanation on how *RemoteAddress,*ReportPort is filled.*

#### HTTP access_log request properties
| metadata's key                    |  example(s)            | description |
|-----------------------------------|------------------------|-------------|
| authority     | `neverssl.com`, `neverssl.com:80` | HTTP/2 `authority` or HTTP/1.1 `Host` header value. |
| forwardedFor  | `203.0.113.195, 70.41.3.18, 150.172.238.178` | X-Forwarded-For HTTP header |
| originalPath  | `/` | Value of the ``X-Envoy-Original-Path`` request header. |
| path          | `/` | The path portion from the incoming request URI. |
| referer       | `Referer: https://www.seznam.cz` | Value of the `Referer` request header. |
| http_*request_header_name* e.g. `http_slo-domain` | `userportal` | Request's HTTP header |
| requestBodyBytes      | `32` ||
| requestHeadersBytes   | `32` ||
| requestId             | `e087fb8b-ee2f-4d92-bb83-afdabc8cceee` | Value of the ``X-Request-Id`` request header |
| requestMethod         | `GET` | HTTP method name |

> | scheme                | `http` | `The scheme portion of the incoming request URI.` |

Reviewer 2: ```suggestion
| scheme                | `http` | The scheme portion of the incoming request URI. |
```

Author: b4b0585

> | userAgent             | `curl/7.74.0-DEV` | `Value of the `User-Agent` request header.` |

Reviewer 2: ```suggestion
| userAgent             | `curl/7.74.0-DEV` | `Value of the `User-Agent` request header. |
```

Author: b4b0585

# Envoy access log server

|                |                        |
|----------------|------------------------|
| `moduleName`   | `envoyAccessLogServer` |
| Module type    | `producer`             |
| Output event   | `raw`                  |

This module allows you to generate events based on access logs sent form remote [Envoy proxy](https://www.envoyproxy.io/) over a gRPC interface.

### Envoy support and configuration
At this moment, V3 of envoy's xDS API is supported. [See the upstream documentation for details](https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versions) on API versions.

In particular, you have to configure your envoy instance to send access_logs using [v3 AccessLogService rpc](https://github.com/envoyproxy/envoy/blob/842485709d651a6057b2ffb505ffced21173e004/api/envoy/service/accesslog/v3/als.proto#L39). We currently do not implement handling of other versions.

See a minimal example on how to configure envoy to send access_logs to an slo-exporter instance (envoy v1.15+):

```yaml
static_resources:

  listeners:

>  	|

Reviewer: please replace tabs by spaces.

Author: Done.

Thanks for catching this.
